packages/transact/src/transact/transact.ts

```diff
@@ -0,0 +1,234 @@
+import {
+  AssetSymbol,
+  Chain,
+  ChainKey,
+  createBalanceBuilder,
+  moonbase,
+  MoonbaseAssets,
+  moonbeam,
+  MoonbeamAssets,
+  moonriver,
+  MoonriverAssets,
+  type MoonbaseChains,
+  type MoonbeamChains,
+  type MoonriverChains,
+  type XcmConfigBuilder,
+} from '@moonbeam-network/xcm-config';
+import {
+  createTxEventHandler,
+  getPolkadotApi,
+  type ExtrinsicEventsCallback,
+  type Hash,
+} from '@moonbeam-network/xcm-utils';
+import type { Signer as PolkadotSigner } from '@polkadot/api/types';
+import type { IKeyringPair } from '@polkadot/types/types';
+import { isString } from '@polkadot/util';
+import type { Signer as EthersSigner } from 'ethers';
+import { XcmTransactorContract } from '../contracts';
+import type {
+  TransactFromData,
+  Transactor,
+  TransactorFrom,
+  TransactorTo,
+  TransactToData,
+} from './transact.interfaces';
+import {
+  getBalance,
+  getDerivatedAddress,
+  getTxWeight,
+  transactThroughSigned,
+} from './transact.utils';
+
+export interface TransactInitOptions {
+  ethersSigner?: EthersSigner;
+  polkadotSigner?: PolkadotSigner;
+}
+
+export function init(options: TransactInitOptions) {
+  return {
+    moonbase: initByChain<MoonbaseAssets, MoonbaseChains>(moonbase, options),
+    moonbeam: initByChain<MoonbeamAssets, MoonbeamChains>(moonbeam, options),
+    moonriver: initByChain<MoonriverAssets, MoonriverChains>(
+      moonriver,
+      options,
+    ),
+  };
+}
+
+function initByChain<Symbols extends AssetSymbol, ChainKeys extends ChainKey>(
+  configBuilder: XcmConfigBuilder<any, ChainKeys>,
+  options?: TransactInitOptions,
+): Transactor<Symbols, ChainKeys> {
+  const { moonChain, transact } = configBuilder;
+  const { chainsFrom, chainsTo, from, to } = transact();
+  const balanceBuilder = createBalanceBuilder<Symbols>();
+  const moonBalanceConfig = balanceBuilder.system();
+
+  return {
+    moonChain,
+    chainsFrom,
+    chainsTo,
+    from: (
+      keyOrChain: ChainKeys | Chain<ChainKeys>,
+    ): TransactorFrom<Symbols, ChainKeys> => {
+      const {
+        balance: balanceConfig,
+        chain,
+        config,
+        getOverallWeight,
+        getOverallFee,
+      } = from(keyOrChain);
+
+      return {
+        transact: async (
+          callHash: string,
+          sourceAccount: string | IKeyringPair,
+          polkadotSigner?: PolkadotSigner,
+        ): Promise<TransactFromData<Symbols, ChainKeys>> => {
+          const signer = polkadotSigner || options?.polkadotSigner;
+
+          if (isString(sourceAccount) && !signer) {
+            throw new Error(
+              'Polkadot Signer/KeyringPair are not provided to xcm-transact',
+            );
+          }
+
+          const sourceAddress = isString(sourceAccount)
+            ? sourceAccount
+            : sourceAccount.address;
+
+          const sourceApi = await getPolkadotApi(chain.ws);
+          const destinationApi = await getPolkadotApi(moonChain.ws);
```

Comment:
I think doing these concurrently might improve things, given how long some endpoints take to connect